scripts/gpg-share.sh

```diff
@@ -125,36 +108,105 @@ check_users() {
 }
 
 
-check_users
-
-DUMP_DIR=$(mktemp -u "/tmp/gpg-share-XXXXXX")
-mkdir -p "$DUMP_DIR"
-
-
 # TODO: documentation
 encrypt () {
-  for _user in "${users[@]}";
+  file="$1"
+  directory="$2"
+  shift 2
+  users=("$@")
+  for user in "${users[@]}";
   do
-      user=$(echo -n "$_user")
-      log_info "encrypting '$file' for '$user' inside '$DUMP_DIR'..."
-      output=$(echo "$DUMP_DIR/$file.$user.asc" | sed 's/\s\+/-/g')
+      log_info "encrypting '$file' for '$user' inside '$directory'..."
+      output=$(echo "$directory/$file.$user.asc" | sed 's/\s\+/-/g')
       gpg --verbose --recipient "$user" --encrypt --armor --output "$output" "$file"
   done
-  log_success "all encrypted file have been saved inside '$DUMP_DIR'"
+  log_success "all encrypted file have been saved inside '$directory'"
 }
 
 
 # TODO: documentation
 archive () {
+  directory="$1"
+  name="$2"
   log_info "storing files in archive"
   # subshell here to avoid archiving the paths to the .asc files
   (
-    cd "$DUMP_DIR"
-    tar cvf "$ARCHIVE" $(find . -type f)
+    cd "$directory"
+    tar cvf "$name" $(find . -type f)
   )
-  cp "$DUMP_DIR/$ARCHIVE" .
+  cp "$directory/$name" .
+}
+
+
+# parse the arguments.
+OPTIONS=$(getopt -o hu:U:f: --long help,user:,users:,file -n 'gpg-share' -- "$@")
+if [ $? != 0 ] ; then echo "Terminating..." >&2 ; exit 1 ; fi
+eval set -- "$OPTIONS"
+
+
+usage () {
+  echo "Usage: $USAGE"
+  echo "Type -h or --help for the full help."
+  exit 0
+}
+
+
+help () {
+  echo "gpg-share:"
+  echo "     a script to share a secret file with others using pgp public keys."
+  echo "     it should guide you when you make mistakes, e.g. providing a non-existing file."
+  echo ""
+  echo "Usage:"
+  echo "     $USAGE"
+  echo ""
+  echo "Options:"
+  echo "     -h/--help               shows this help."
+  echo "     -f/--file               the secret file to share."
+  echo "     -u/--user               a single user id in the keyring."
+  echo "                               can be used multiple times."
+  echo "                               supports multi-word ids inside quotes, e.g. \"John Smith\"."
+  echo "     -U/--users              the path to a file containing pgp ids."
+  echo "                               should contain exactly one valid pgp id per line."
+  echo "                               can be used multiple times."
+  exit 0
+}
+
+
+main () {
+  users=()
+  while [[ $# -gt 0 ]]; do
+    case "$1" in
+      -h | --help ) help ;;
+      -f | --file ) file="$2"; shift 2;;
+      -u | --user ) users+=("$2"); shift 2;;
+      -U | --users ) 
+        check_file "$2"
+        readarray new_users < "$2"
+        shift 2
+        for new_user in "${new_users[@]}"; do
+          users+=("$(echo -n "$new_user")")
+        done;;
+      -- ) shift; break ;;
+      * ) break ;;
+    esac
+  done
+
+  check_dependencies
+  check_pwd
+
+  check_file "$file"
```

Comment:
Maybe we should do this in arg parsing as for the `-U` option!

Reply:
i am used to parsing all the arguments and then running the actual script :thinking: 

i think it's better to separate the two to avoid subtle bugs :confused:

Reply:
I get your point :yum: just wanted to have symmetrical code between `-U` handling and `-f` (even if I get that you're **forced** to do it the `switch case` for `-U`) 
Feel free to resolve this if you want to stand your ground :+1:

Reply:
> I get your point yum just wanted to have symmetrical code between `-U` handling and `-f` (even if I get that you're **forced** to do it the `switch case` for `-U`)
>
yup, i think it's far better and easier to build the array of users directly in the parsing.
this is a very low overhead to the parsing of the arguments and imo the goal of parsing the arguments here is to 
- have the filename
- have the full array of users

> Feel free to resolve this if you want to stand your ground +1
>
i'll let the parsing as it is then :relieved: